### ISSUE TITLE
[ray] Why not separate RLlib/Serv/SGD/Tune from ray

### ISSUE THREAD

**Issue by DonYum** [ray-project/ray#12747]
There's definitely a lot of thought and discussions about this topic. Can anybody show me some information?
Any response will be appreciated.

Thx and BRs!

**Comment by krfricke** [comment]
Hey @DonYum, there are a couple of benefits and drawbacks with our monorepo approach vs. keeping the libraries in separate repos. One of the most daunting things here is cross-compatibility. E.g. RLLib integrates nicely with Tune (you can call tune.run directly on RLLib Trainers). Tune on the other hand auto-detects autoscaling environments and configures syncers accordingly. 

If we kept these libraries in different repositories, it would be much headache to make sure they work together smoothly. If we want to implement a new feature or change a API, we would potentially have to create a PR in three repositories.

With the monorepo approach we can at least make sure that each commit to the master branch keeps compatibility between these libraries.

Note that we have a couple of external repos (e.g. for [tune-sklearn](https://github.com/ray-project/tune-sklearn) and [xgboost_ray](https://github.com/ray-project/xgboost_ray) where keeping compatibility sometimes is a headache already, but since we keep one-way dependencies there it's not as bad.

**Comment by DonYum** [comment]
@krfricke  Your answer solved my confusion perfectly.
Thx a lot!

**Comment by DonYum** [comment]
@krfricke I have another question now: why not detach ray core from this repo?
As you say, there are much cross-compatibility to consider. But ray core is a basic 'toolbox' which should provide enough APIs.
If we achive that, ray core should be separated from other projects.

The reason I'm asking this question is because I see that ray is now too big.
If this continues, maybe, it will seriously slow down development progress.

**Comment by krfricke** [comment]
Currently we'll be keeping Ray as a monorepo, as changes in some of the ray core functionality might impact the libraries (though not vice versa). The overhead from carrying the libraries should be relatively low: It's just text files of python code, e.g. no modules are compiled, and dependencies are only installed if requested via `pip install ray[all]`, or `ray[rllib]`, `ray[tune]` etc. We're aware of the potential pitfalls however.

Thanks for your input, we'll keep it in mind! I'll close this issue here, but if you'd like to discuss this more thoroughly please feel free to open a post at https://discuss.ray.io/!

**Comment by DonYum** [comment]
OK, thanks for your patient reply.
BRs!
